libyul/backends/evm/EVMCodeTransform.cpp

```diff
@@ -628,13 +680,55 @@ void CodeTransform::visitExpression(Expression const& _expression)
 	expectDeposit(1, height);
 }
 
+void CodeTransform::allocateReturnSlotsIfNeeded(Statement const& _statement)
+{
+	if (m_returnStackHeight.has_value())
+		return;
+	if (holds_alternative<FunctionDefinition>(_statement))
+		return;
+	if (
+		holds_alternative<ExpressionStatement>(_statement) ||
+		holds_alternative<Assignment>(_statement)
+	)
+	{
+		ReferencesCounter referencesCounter{ReferencesCounter::CountWhat::OnlyVariables};
+		referencesCounter.visit(_statement);
+		auto isReferenced = [&](YulString _returnVariableName) {
+			return referencesCounter.references().count(_returnVariableName);
+		};
+		if (ranges::none_of(m_delayedReturnVariables, isReferenced))
+			return;
+	}
+
+	for (YulString _returnVariableName: m_delayedReturnVariables)
+	{
+		auto* var = std::get_if<Scope::Variable>(m_scope->lookup(_returnVariableName));
+		yulAssert(var, "Return variable not in scope.");
+		m_context->variableStackHeights[var] = static_cast<size_t>(m_assembly.stackHeight());
```

Comment:
Why don't you re-use unused stack slots?

Reply:
Because it complicates things further and is not strictly necessary for the allowing the stack evasion transformation :-).

It did cause problems at some point, but it might be fine now, I could try and see what happens, if I do try reusing them again in the latest version. Might be fine.

Although it might move the return slot too deep into the stack...

E.g.
```
function f(a1, ..., a15, a16) -> x {
  sstore(0, a16) // last use of a6
  x := 42 // reuses slot of a16
  let b1, ..., b15 := g()
  x := 21 // x has become unreachable
  h(b1, ..., b15) // the b's are used somehow and can't be freed
  h(a1, ..., a15) // the other arguments are used somehow and can't be freed
}
```
but right now I'm not sure if the opposite, i.e. not reusing the slot is always better. It's complicated in any case :-).

Reply:
The deeper down return variables are the better, isn't it? They are the last ones that will always remain used.